README.md
# Technical Test: Microservice Project

## Description

This document will outline the requirements, tools, and technologies I utilized to meet the requested criteria.

The reproduction of this project should only be for evaluation and feedback purposes.

## Whoami

I am a sysadmin with over:

- 3 years administering Kubernetes.
- 5 years in Docker administration.
- 5 years of advanced GIT knowledge.
- 20 years of advanced GNU/Linux administration.
- Proficient in: 3 years with Python, 1 year with Java, Go, Ruby, and Node.
- 5 years knowledgeable in CI/CD pipelines, on both local platforms and in the cloud.
- 7 years in Cloud Computing management with Microsoft Azure, AWS, Google Cloud Platform, OCI.
- 5 years of experience in IaC (Infrastructure as Code), with traditional providers, as well as OCI, VMware, and Digital Ocean.

## Schema

Constructed based on technology, services, and goals is a directory tree, which contains the current repository project.

![schema][schema]

## Requirements

You should install the following tools on your local machine:

- ngrok.
- Python.
- Docker Desktop.

A microservice is constructed containing a REST EndPoint called: `/DevOps`.

- We use Python as the programming language and Flask as the Framework.
- A script from the project is used to generate a token, in case the current JWT has expired:

```bash
python docker/python/Scripting/generate_jwt.py
```

- We use the HTTP POST method on the JSON file to utilize the Payload with the required API Key, validating all request fields:

```bash
curl -X POST \
-H "X-Parse-REST-API-Key: 2f5ae96c-b558-4c7b-a590-a501ae1c3f6c" \
-H "X-JWT-KEY: $JWT_Token" \
-H "Content-Type: application/json" \
-d '{"message": "This is a test", "to": "Juan Perez", "from": "Rita Asturia", "timeToLifeSec": 45}' \
http://$IP_ADDRESS_POD:$PORT/DevOps
```

Having the following output:

```bash
{
  "message": "Hello Juan Perez your message will be send"
}
```

- If we use another HTTP method:

```bash
curl -X POST \
-H "X-Parse-REST-API-Key: 2f5ae96c-b558-4c7b-a590-a501ae1c3f6c" \
-H "X-JWT-KWY: $JWT_Token" \
-H "Content-Type: application/json" \
-d '{"message": "This is a test", "to": "Juan Perez", "from": "Rita Asturia", "timeToLifeSec": 45}' \
https://ecf7-$NGROK-Public-Address-IP.ngrok-free.app/DevOps
```

It will return "ERROR":

```bash
ERROR
```

Within the microservice running in the container, it returns:

```bash
 * Serving Flask app 'devops'
 * Debug mode: on
WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:5000
 * Running on http://172.17.0.2:5000
Press CTRL+C to quit
 * Restarting with stat
 * Debugger is active!
 * Debugger PIN: 123-062-598
172.17.0.1 - - [21/Oct/2023 08:54:02] "POST /DevOps HTTP/1.1" 200 -
172.17.0.1 - - [21/Oct/2023 08:55:43] "GET /DevOps HTTP/1.1" 405 -
172.17.0.1 - - [21/Oct/2023 08:56:13] "PUT /DevOps HTTP/1.1" 405 -
172.17.0.1 - - [21/Oct/2023 08:56:32] "DELETE /DevOps HTTP/1.1" 405 -
172.17.0.1 - - [21/Oct/2023 08:56:46] "PATCH /DevOps HTTP/1.1" 405 -
172.17.0.1 - - [21/Oct/2023 08:56:56] "POST /DevOps HTTP/1.1" 400 -
172.17.0.1 - - [21/Oct/2023 08:58:10] "POST /DevOps HTTP/1.1" 400 -
172.17.0.1 - - [21/Oct/2023 08:59:45] "POST /DevOps HTTP/1.1" 200 -
```

## How to run

We use the necessary context for Kubernetes:

```bash
minikube stop
minikube start
kubectl config use-context docker-desktop
kubectl config get-contexts
```

We set up the ports using a service:

```bash
kubectl apply -f .\kubernetes\microservice-service.yml
kubectl get services
```

We deploy on Kubernetes for load balancing across 2 pods:

```bash
kubectl describe deployment microservice-deployment
kubectl get deployments
kubectl get pods
```

## Updates

- **22-10-2023**: Refactoring and redesign the project for Unit Test. Reviewed by [Oscar Macias]![chat].

## Support Contact

For questions or issues, please contact on [LinkedIn]![social] or by [e-mail]![mail]. Also, check out our [Python Wiki]![wiki].

[schema]: ./src/img/schema.png

**[chat]: <https://wa.me/573058288031>**

**[mail]: <mailto:omaciasnarvaez@gmail.com>**

**[social]: <https://www.linkedin.com/in/omaciasd/>**

**[wiki]: <https://github.com/OMaciasd/Microservices>**


CHANGELOG.md
# History

- **20-10-2023**: Implementacion  en  "Dependencias del microservicio en Python".  Revisado por [Oscar Macias]![chat].

- **21-10-2023**: Despliegue de microservicio contenerizado por Docker, balanceado por Kubernetes.  Revisado por [Oscar Macias]![chat].

- **22-10-2023**: Refactoring and redesign the project for Unit Test. Reviewed by [Oscar Macias]![chat].


tests/test_app.py
import unittest
import sys
import os

base_dir = os.path.dirname(os.path.abspath(__file__))
app_path = os.path.join(base_dir, '..', 'app')
if app_path not in sys.path:
    sys.path.append(app_path)

from app import create_app

app = create_app('testing')

class TestDevOpsApp(unittest.TestCase):

    def setUp(self):
        self.app = app
        self.app.testing = True
        self.client = self.app.test_client()
        response_jwt = self.client.get('/generate-jwt')

        print(response_jwt.status_code)

        json_data = response_jwt.get_json()
        if json_data:
            self.valid_jwt = json_data.get("token", None)
        else:
            self.valid_jwt = None

    def post_to_devops(self, headers, data=None):
        return self.client.post('/devops', headers=headers, json=data)

    def test_generate_jwt(self):
        response = self.client.get('/generate-jwt')
        self.assertEqual(response.status_code, 200)

        print(response.data)

    # def test_devops_endpoint_post(self):
    #     headers = {
    #         'X-Parse-REST-API-Key': '2f5ae96c-b558-4c7b-a590-a501ae1c3f6c',
    #         'X-JWT-KWY': self.valid_jwt,
    #         'Content-Type': 'application/json'
    #     }
    #     data = {
    #         'message': 'Test Message',
    #         'to': 'Test Receiver',
    #         'from': 'Test Sender',
    #         'timeToLifeSec': 30
    #     }
    #     response = self.post_to_devops(headers, data)  # Usar el método de apoyo

    #     # Check HTTP status code
    #     self.assertEqual(response.status_code, 200)

    #     # Check response content
    #     json_data = response.get_json()
    #     self.assertEqual(json_data["message"], "Hello Test Receiver your message will be send")

    # def test_devops_endpoint_get(self):
    #     response = self.client.get('/DevOps')
    #     self.assertEqual(response.data, b'ERROR')
    #     self.assertEqual(response.status_code, 405)

    # def test_invalid_api_key(self):
    #     headers = {
    #         'X-Parse-REST-API-Key': 'invalid_api_key',
    #         'X-JWT-KWY': self.valid_jwt,
    #         'Content-Type': 'application/json'
    #     }
    #     response = self.post_to_devops(headers)  # Usar el método de apoyo
    #     self.assertEqual(response.status_code, 401)

    # def test_invalid_jwt(self):
    #     headers = {
    #         'X-Parse-REST-API-Key': '2f5ae96c-b558-4c7b-a590-a501ae1c3f6c',
    #         'X-JWT-KWY': 'invalid_jwt_token',
    #         'Content-Type': 'application/json'
    #     }
    #     response = self.post_to_devops(headers)
    #     self.assertEqual(response.status_code, 401)

    # def test_incomplete_payload(self):
    #     headers = {
    #         'X-Parse-REST-API-Key': '2f5ae96c-b558-4c7b-a590-a501ae1c3f6c',
    #         'X-JWT-KWY': self.valid_jwt,
    #         'Content-Type': 'application/json'
    #     }
    #     data = {'message': 'Test Message'}  # Payload incompleto
    #     response = self.client.post('/DevOps', headers=headers, json=data)
    #     self.assertEqual(response.status_code, 400)

    # def test_devops_endpoint_put(self):
    #     headers = {
    #         'X-Parse-REST-API-Key': '2f5ae96c-b558-4c7b-a590-a501ae1c3f6c',
    #         'X-JWT-KWY': self.valid_jwt
    #     }
    #     response = self.client.put('/DevOps', headers=headers)
    #     self.assertEqual(response.data.decode("utf-8"), "ERROR")
    #     self.assertEqual(response.status_code, 405)

if __name__ == "__main__":
    unittest.main()


docker/python/devops.py
from flask import Flask, request, jsonify
import jwt
import uuid
import datetime

app = Flask(__name__)

API_KEY = "2f5ae96c-b558-4c7b-a590-a501ae1c3f6c"
SECRET_KEY = "supersecret"

@app.route('/generate-jwt', methods=['GET'])
def generate_jwt_endpoint():
    payload = {
        "user": "example_user",
        "exp": datetime.datetime.utcnow() + datetime.timedelta(seconds=3600),
        "transaction_id": str(uuid.uuid4())  # Genera un UUID para cada transacción
    }
    encoded_jwt = jwt.encode(payload, SECRET_KEY, algorithm="HS256")
    return jsonify({"jwt": encoded_jwt})

@app.route('/DevOps', methods=['POST', 'GET', 'PUT', 'DELETE', 'PATCH'])
def devops_endpoint():
    if request.method != 'POST':
        return "ERROR", 405

    received_api_key = request.headers.get("X-Parse-REST-API-Key")
    if not received_api_key == API_KEY:
        return "ERROR: Invalid API Key", 401

    jwt_header = request.headers.get("X-JWT-KWY")
    if not jwt_header:
        return "ERROR: Missing JWT Header", 400

    # Decodificar y validar JWT
    try:
        decoded_payload = jwt.decode(jwt_header, SECRET_KEY, algorithms=["HS256"])
    except jwt.ExpiredSignatureError:
        return "ERROR: JWT Token has expired", 401
    except jwt.InvalidTokenError:
        return "ERROR: Invalid JWT Token", 401

    # Aquí verificamos el usuario en el JWT
    if decoded_payload["user"] != "example_user":
        return "ERROR: Unauthorized user", 401

    data = request.json
    if not all(field in data for field in ["message", "to", "from", "timeToLifeSec"]):
        return jsonify({"error": "Incorrect Payload"}), 400

    return jsonify({
        "message": f"Hello {data['to']} your message will be send"
    }), 200

if __name__ == '__main__':
    app.run(host='0.0.0.0', debug=True)


app/views.py
from flask import Blueprint, request, jsonify
import jwt
import os
import pathlib

env_path = pathlib.Path(__file__).parent.parent / 'config' / '.env'

with open(env_path, 'r') as f:
    lines = f.readlines()
    for line in lines:
        key, value = line.strip().split('=', 1)
        os.environ[key] = value

API_KEY = os.environ.get('API_KEY')
SECRET_KEY = os.environ.get('SECRET_KEY')

if not API_KEY:
    raise ValueError("API_KEY is not set.")
if not SECRET_KEY:
    raise ValueError("SECRET_KEY is not set.")

main = Blueprint('main', __name__)

@main.route('/generate-jwt', methods=['GET'])
def generate_jwt_endpoint():
    payload = {"user": "example_user"}
    token = jwt.encode(payload, SECRET_KEY, algorithm="HS256")
    return jsonify({"token": token})

@main.route('/DevOps', methods=['POST', 'GET', 'PUT', 'DELETE', 'PATCH'])
def devops_endpoint():
    if request.method != 'POST':
        return "ERROR", 405

    received_api_key = request.headers.get("X-Parse-REST-API-Key")
    print(f"Received API Key: {received_api_key}")

    if not received_api_key == API_KEY:
        return "ERROR: Invalid API Key", 401

    jwt_header = request.headers.get("X-JWT-KWY")
    if not jwt_header:
        return "ERROR: Missing JWT Header", 400

    try:
        decoded_payload = jwt.decode(jwt_header, SECRET_KEY, algorithms=["HS256"])
    except jwt.ExpiredSignatureError:
        return "ERROR: JWT Token has expired", 401
    except jwt.InvalidTokenError:
        return "ERROR: Invalid JWT Token", 401

    if decoded_payload["user"] != "example_user":
        return "ERROR: Unauthorized user", 401

    data = request.json
    if not all(field in data for field in ["message", "to", "from", "timeToLifeSec"]):
        return jsonify({"error": "Incorrect Payload"}), 400

    return jsonify({
        "message": f"Hello {data['to']} your message will be send"
    }), 200
